File: Python/MemoryDatabase.py


#@author Kowshick Srinivasan
#@version: 1.0
#@Assignment: Hw2
import csv
import os
import threading

from LinkedList import LinkedList
from Student import Student


def convert_bool(param):
    if param == "yes":
        return True
    else:
        return False


def convert_to_student(row):
    line = list(row)
    return Student(line[0], line[1], int(line[2]), line[3], line[4], line[5],
                   int(line[6]), int(line[7]), line[8], line[9], line[10], line[11],
                   int(line[12]), int(line[13]), int(line[14]), convert_bool(line[15]), convert_bool(line[16]),
                   convert_bool(line[17]), convert_bool(line[18]), convert_bool(line[19]),
                   convert_bool(line[20]), convert_bool(line[21]), convert_bool(line[22]), int(line[23]), int(line[24]),
                   int(line[25]), int(line[26]),
                   int(line[27]), int(line[28]), int(line[29]), convert_bool(line[30]))


# Memory database class that does:
# a) load the student-data. csv file into the memory database through recursive function.
# b) When a row is added or removed from the CSV file, the memory database will adapt to the changes.
# c) export the memory database into a csv file through recursive function.
class MemoryDatabase:
    BASEPATH = r"D:\Algorithms\IndHw2\Data"
    FILENAME = os.path.join(BASEPATH, 'student-data.csv')
    OUTFILE = os.path.join(BASEPATH, 'student-op-data.csv')

    def read(self, current_node, input_csv):
        """
        Recursive method to read the CSV file and add the contends to the back of the Linked list
        :param current_node: The current Node
        :param input_csv: Reader object
        :return: Completion status of the read operation
        """
        row = next(input_csv, None)
        if row is None:
            return 0
        student = convert_to_student(row)
        return self.read(self.list.insert(current_node, student, None), input_csv)

    # Constructor based dependency injection
    def __init__(self):
        self.list = LinkedList()
        self.running = True
        self.prev_node = None

    def run(self):
        with open(self.FILENAME, newline="") as inputFile:
            input_csv = csv.reader(inputFile)
            header = next(input_csv, None)  # read the Header of the csv
            self.read(self.list.head, input_csv)  # Read the student details from the file and add it to the linked list
        print('enter 0 to stop watching')  # KeyBoard interrupt to signal make watcher thread to stop and join back to the main
        watcher_thread = threading.Thread(target=self.file_watcher)  # Load the watcher method to the watcherThread
        watcher_thread.start()  # Start the watcher thread

        if input() == '0':  # Blocking-instruction, makes the main thread to wait until there is user input 0
            self.running = False  # Signal the watcher thread to stop

        watcher_thread.join()  # Resume after the completion of watcher thread

        with open(self.OUTFILE, mode='w', newline="") as outputFile:
            output_csv_writer = csv.writer(outputFile)
            output_csv_writer.writerow(header)  # print the Header of the csv
            self.printer(self.list.head,
                         output_csv_writer)  # Method to print the traverse the Linked list and print in the csv

    def printer(self, current, output_csv_writer):
        """
        Recursively prints the student data to the csv file
        :param current: Accepts the current NOde/ pointer of the Linked List
        :param output_csv_writer: Accepts the Printer writer object
        :return: completion status of the printing action
        """
        row = current.student.__str__()
        output_csv_writer.writerow(row)
        if current.next is None:
            return 0
        return self.printer(current.next, output_csv_writer)

    # Watches for any changes in the csv file
    def file_watcher(self):
        last_modified = None
        if os.path.exists(self.FILENAME):
            last_modified = os.path.getmtime(self.FILENAME)
        while self.running:  # run the loop until the condition variable is true
            if os.path.exists(self.FILENAME):
                current_modified = os.path.getmtime(self.FILENAME)
                if last_modified is None or current_modified != last_modified:
                    print("changes detected in file")
                    with open(self.FILENAME, mode='r', newline='') as modified_file:
                        modified_csv = csv.reader(modified_file)
                        next(modified_csv, None)  # skip header
                        self.reload(self.list.head, modified_csv)  # Reload only the changes
                        last_modified = current_modified

    def reload(self, current_node, input_csv):
        row = next(input_csv, None)
        # When both the linked list and file is not at the end,
        # i.e. insertion/deletion happens at head in the middle
        if current_node.next is not None and row is not None:
            new_student = convert_to_student(row)
            # There is a change in the current node and the file row, investigate further
            if not new_student.__eq__(current_node.student):  #######problem
                next_node = current_node.next
                # if the current node is the last or if the next node is same as the current row,
                # then the current row has been deleted, so we need to delete current node
                if next_node.next is None or next_node.student.__eq__(new_student):
                    self.list.delete(self.prev_node, next_node)  # deletion operation
                else:
                    self.list.insert(self.prev_node, new_student, next_node)  # Insertion operation

            self.prev_node = current_node  # Store the current node
            return self.reload(current_node.next, input_csv)  # recursively call with the next node and reader object
        # If the linked list has reached its end but not the file,
        # then there is something to be inserted at the end of the Linked list
        elif current_node.next is None and row is not None:
            new_student = convert_to_student(row)
            new_node = self.list.insert(self.prev_node, new_student, None)
            self.prev_node = new_node  # Store the current node
            return self.reload(new_node, input_csv)  # recursively call with the next node and reader object
        # If the File has reached its end and not the Linked List,
        # there is something to be deleted from the end of the LinkedList
        elif current_node.next is not None:
            self.list.delete(self.prev_node, None)
            # No need to recursively call since we need to delete everything that are excess
        return 0  # both File and Linked list has reached the end


if __name__ == '__main__':
    db = MemoryDatabase()
    db.run()

File: Java/MemoryDatabase.java
package org.example;

import java.io.*;
import java.nio.file.*;
import java.util.Objects;
import java.util.Scanner;
/*
  @author: Kowshick Srinivasan
 * @version: 1.0
 * @Assignment: Hw2
 */

/**
 * Memory database class that does:
 * <p>
 * a) load the student-data.csv file into the memory database through recursive function.
 * <p>
 * b) When a row is added or removed from the CSV file, the memory database will adapt to the changes.
 * <p>
 * c) export the memory database into a csv file through recursive function.
 */
public class MemoryDatabase {

    LinkedList list;  //Object of the linked list

    private static final String FILENAME = "student-data.csv";  //input file name

    private static final String OUTFILE = "student-op-data.csv"; //output file name

    LinkedList.Node prevNode;

    Scanner sc;  //Scanner object

    private static boolean running = true;  //Condition variable

    //Constructor based dependency injection
    public MemoryDatabase(LinkedList list, Scanner sc, LinkedList.Node prevNode) {
        this.list = list;
        this.sc = sc;
        this.prevNode = prevNode;
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);
        MemoryDatabase database = new MemoryDatabase(
                new LinkedList()  //Initialize the list object
                , new Scanner(System.in), null); //Initialize the scanner object

        //Since we used try with resources, we won't be needing a finally block
        try (BufferedReader br = new BufferedReader(new FileReader(FILENAME)); //Buffered reader to enable to the Java methods to read from the given file
             PrintWriter printWriter = new PrintWriter(new FileWriter(OUTFILE))) //Buffered writer to enable the Java methods to write on the given file
        {
            String header = br.readLine();  //read the Header of the csv
            database.read(br, database.list.head);   //Read the student details from the file and add it to the linked list
            System.out.println("Press 0 to exit");   //KeyBoard interrupt to signal make watcher thread to stop and join back to the main
            Thread watcherThread = new Thread(() -> {  //Load the watcher method to the watcherThread
                try {
                    database.watcher();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            });
            watcherThread.start();  //Start the watcher thread
            while (true) {
                if (scanner.hasNext() && Objects.equals(scanner.nextLine(), "0")) {  //If there is any keyboard input, and it is 0
                    running = false; //change the contention variable to false
                    //watcherThread.join();   //Resume after the completion of watcher thread
                    System.out.println("Stopping watcher..."); //Stop the watcher service
                    break;
//                    }
                }
            }


            printWriter.println(header);  //print the Header of the csv
            database.print(database.list.head, printWriter);  //Method to print the traverse the Linked list and print in the csv


        } catch (Exception e) { //Global catch block to handle all the exception thrown by the program
            e.printStackTrace(System.out);  //print the exception stack trace in console

        }

    }


    /**
     * Method recursively read each student details from the CSV file and adds it to the end of the linked list
     *
     * @param br      : The BufferedReader object
     * @param current : Current Node/pointer of the linked list
     * @return : Completion status of the read operation
     */
    private int read(BufferedReader br, LinkedList.Node current) throws IOException {
        String l = br.readLine();  //Read each line
        if (l == null) return 0;  //When reached EOF return
        String[] line = l.split(",");  //Split the row using "," as the separator
        Student student = convertLineToStudent(line);  //Convert the String array to the student class
        return read(br, list.insert(current, student, null));   //Recursively call the function with the buffered reader object and updated last Node/Pointer
    }


    /**
     * Accepts the String array and convert it to student object
     *
     * @param line Original String array formed after comma separation of the line
     * @return Student object
     */
    private Student convertLineToStudent(String[] line) {
        return new Student(line[0], line[1].charAt(0), Integer.parseInt(line[2]), line[3].charAt(0), line[4], line[5].charAt(0),
                Integer.parseInt(line[6]), Integer.parseInt(line[7]), line[8], line[9], line[10], line[11],
                Integer.parseInt(line[12]), Integer.parseInt(line[13]), Integer.parseInt(line[14]), line[15].equals("yes"), line[16].equals("yes"), line[17].equals("yes"), line[18].equals("yes"), line[19].equals("yes"),
                line[20].equals("yes"), line[21].equals("yes"), line[22].equals("yes"), Integer.parseInt(line[23]), Integer.parseInt(line[24]), Integer.parseInt(line[25]), Integer.parseInt(line[26]),
                Integer.parseInt(line[27]), Integer.parseInt(line[28]), Integer.parseInt(line[29]), line[30].equals("yes"));

    }


    /**
     * Recursively prints the student data to the csv file
     *
     * @param current     : Accepts the current NOde/pointer of the Linked List
     * @param printWriter : Accepts the Printed writer object
     * @return : return the completion status of the printing action
     */
    private int print(LinkedList.Node current, PrintWriter printWriter) {
        String row = current.student.toString(); //Converts the student object to comma separated string
        printWriter.println(row);  //Prints the string to the csv file
        if (current.next == null)  //When reached end node return completion status
            return 0;
        return print(current.next, printWriter); //recursively call the print method with updated last node pointer
    }

    /**
     * Watches for any changes in the csv file
     */
    private void watcher() throws Exception {
        WatchService watchService = FileSystems.getDefault().newWatchService(); //create a watcher service
        Path path = Paths.get("."); // current directory
        path.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);  //Notify when there is a modification event

        while (running) {  //run the loop until the condition variable is true
            WatchKey key = watchService.take(); // blocks until an event
            for (WatchEvent<?> event : key.pollEvents()) {
                WatchEvent.Kind<?> kind = event.kind();
                Path changed = (Path) event.context();
                if (changed.endsWith(FILENAME) && kind == StandardWatchEventKinds.ENTRY_MODIFY) {
                    System.out.println("student-data.csv has changed!");  //Notify there is a change detected in the file
                    BufferedReader br = new BufferedReader(new FileReader(FILENAME)); //buffered reader object of the file
                    br.readLine();  //skip the header
                    reload(list.head, br);  //Reload only the changes

                }

            }
            boolean valid = key.reset();
            if (!valid) break;
        }
    }


    private int reload(LinkedList.Node current, BufferedReader br) throws IOException {
        String row = br.readLine();
        if (current.next != null && row != null) {  //When both the linked list and file is not at the end,
            // i.e. insertion/deletion happens at head in the middle
            String[] line = row.split(",");
            Student newStudent = convertLineToStudent(line);
            //Change detected
            if (!newStudent.equals(current.student)) {  //There is a change in the current node and the file row, investigate further
                LinkedList.Node nextNode = current.next;
                if (nextNode.next == null || nextNode.student.equals(newStudent)) {   //if the current node is the last or if the next node is same as the current row,
                    // then the current row has been deleted, so we need to delete current node
                    list.delete(prevNode, nextNode); //deletion operation
                } else
                    list.insert(prevNode, newStudent, nextNode); //Insertion operation
            }
            prevNode = current;  //Store the current node
            return reload(current.next, br);  //recursively call with the next node and reader object
        } else if (current.next == null && row != null) {   //If the linked list has reached its end but not the file,
            // then there is something to be inserted at the end of the Linked list
            String[] line = row.split(",");
            Student newStudent = convertLineToStudent(line);
            LinkedList.Node newNode = list.insert(prevNode, newStudent, null);
            prevNode = newNode; //Store the current node
            return reload(newNode, br);  //recursively call with the next node and reader object
        } else if (current.next != null) {   //If the File has reached its end and not the Linked List,
            // there is something to be deleted from the end of the LinkedList
            list.delete(prevNode, null);  //delete all the Node that are excess than the File
            //No need to recursively call since we need to delete everything that are excess
        }
        return 0;     //both File and Linked list has reached the end
    }
}

